### ISSUE TITLE
dotnet ef database migrations cause "Unable to connect" exception

### ISSUE THREAD

**Issue by DMWOO7** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#279]
**UPDATE: Found out that I had a bit of misunderstanding about the responsibility of a Dockerfile. Full networking is not avaliable when a container is build. In short, best practice seems to be apply migrations in a container, which got removed later. I want to write a bit more about this and would give later a more detailed answear.**

### Steps to reproduce
Creating a empty ASP.NET Core project, add the packages like in the following csproj file

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp1.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore" Version="1.1.2" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="1.1.2" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="1.1.2" />
    <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="1.1.2" />
  </ItemGroup>

</Project>
```
Now add a simple DbContext with a entity for testing purpose: 

```
public class AppDbContext : DbContext {
        public DbSet<TestEntity> TestEntitys { get; set; }

        public AppDbContext(DbContextOptions options) : base(options) { }
    }

    public class TestEntity {
        public int Id { get; set; }
        public string Name { get; set; }
    }
```

Register the context with Pomelo.EntityFrameworkCore.MySql driver in `ConfigureServices`

    public class Startup {
        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services) {
            string connectionString = "Server=mariadb; Database=test; Uid=root; Pwd=test;";
            services.AddDbContext<AppDbContext>(options => options.UseMySql(connectionString));
        }
}

Now deploy the project on a Linux box (I used Ubuntu 16.10) for deploying them in a docker image. I used the following simple docker file for demonstrating purpose:

```
FROM microsoft/aspnetcore-build:1.1.1
env APP_NAME Testapp
ADD . /app
WORKDIR /app/$APP_NAME

RUN rm -rf /published \
        && mkdir /published \
        && dotnet restore \
        && dotnet ef database update\
        && dotnet publish --output /published/

WORKDIR /published/
ENTRYPOINT dotnet ${APP_NAME}.dll
```
At least try to build a test image based on this docker file: `docker build . -t test1`
Note, that you need to have a named MariaDb Container `mariadb`. In my productive app, I made this on docker compose (mariadb:latest) and link them via depends: 

myapp: 
    depends_on:
        - mariadb

Then you can include the Dockerfile as `build` entry and build it using `docker compose build myapp`. 

### The issue
When applying the migration, I got an exception `MySql.Data.MySqlClient.MySqlException: Unable to connect to any of the specified MySQL hosts`. This is completely not understandable for me, since I double checked the ConnectionString. It's working outside the Docker container via VPN on Windows 10 64 Bit. And for testing purpose, I installed the mysql client in the Docker container which runs the ASP.NET Core app. I was able to connect to the MariaDB server, so it's not a network issue. 

** Full Exception and Stack Trace**
```
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.59
MySql.Data.MySqlClient.MySqlException: Unable to connect to any of the specified MySQL hosts.
   at MySql.Data.Serialization.MySqlSession.<ConnectAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<CreateSessionAsync>d__62.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlConnectionSettings.<>c__DisplayClass2_0.<GetSettings>b__0(String key)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlScopedTypeMapper.get_ConnectionSettings()
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlScopedTypeMapper.MaybeConvertMapping(RelationalTypeMapping mapping)
   at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.IsTypeMapped(IRelationalTypeMapper typeMapper, Type clrType)
   at System.Linq.Enumerable.WhereArrayIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.PropertyDiscoveryConvention.Apply(InternalEntityTypeBuilder entityTypeBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnEntityTypeAdded(InternalEntityTypeBuilder entityTypeBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model.AddEntityType(EntityType entityType, Boolean runConventions)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalModelBuilder.Entity(Type type, ConfigurationSource configurationSource, Boolean runConventions)
   at Microsoft.EntityFrameworkCore.ModelBuilder.Entity(Type type)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.FindSets(ModelBuilder modelBuilder, DbContext context)
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.RelationalModelSource.FindSets(ModelBuilder modelBuilder, DbContext context)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.LazyRef`1.get_Value()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServiceCollectionExtensions.<>c.<AddEntityFramework>b__0_13(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalServiceCollectionExtensions.<>c.<AddRelational>b__0_8(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_1.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Unable to connect to any of the specified MySQL hosts.
The command '/bin/sh -c rm -rf /published         && mkdir /published         && dotnet restore         && dotnet ef database update        && dotnet publish --output /published/' returned a non-zero code: 1
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.59
MySql.Data.MySqlClient.MySqlException: Unable to connect to any of the specified MySQL hosts.
   at MySql.Data.Serialization.MySqlSession.<ConnectAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<CreateSessionAsync>d__62.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlConnectionSettings.<>c__DisplayClass2_0.<GetSettings>b__0(String key)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlScopedTypeMapper.get_ConnectionSettings()
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlScopedTypeMapper.MaybeConvertMapping(RelationalTypeMapping mapping)
   at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.IsTypeMapped(IRelationalTypeMapper typeMapper, Type clrType)
   at System.Linq.Enumerable.WhereArrayIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.PropertyDiscoveryConvention.Apply(InternalEntityTypeBuilder entityTypeBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnEntityTypeAdded(InternalEntityTypeBuilder entityTypeBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model.AddEntityType(EntityType entityType, Boolean runConventions)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.InternalModelBuilder.Entity(Type type, ConfigurationSource configurationSource, Boolean runConventions)
   at Microsoft.EntityFrameworkCore.ModelBuilder.Entity(Type type)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.FindSets(ModelBuilder modelBuilder, DbContext context)
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.RelationalModelSource.FindSets(ModelBuilder modelBuilder, DbContext context)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.LazyRef`1.get_Value()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServiceCollectionExtensions.<>c.<AddEntityFramework>b__0_13(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalServiceCollectionExtensions.<>c.<AddRelational>b__0_8(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass16_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_1.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Unable to connect to any of the specified MySQL hosts.
The command '/bin/sh -c rm -rf /published         && mkdir /published         && dotnet restore         && dotnet ef database update        && dotnet publish --output /published/' returned a non-zero code: 1

```

### Further technical details

MySQL version: MariaDb 10.1.23
Operating system: Ubuntu 16.10 with Docker
Pomelo.EntityFrameworkCore.MySql version: 1.1.2

**Comment by caleblloyd** [comment]
This is most likely a network issue. Install MySQL client and get it to connect from the same environment. Tcpdump and Wireshark are your friends to debug stuff like this.

**Comment by DMWOO7** [comment]
Found out that I had a bit of misunderstanding about the responsibility of a Dockerfile. Full networking is not avaliable when a container is build. In short, best practice seems to be apply migrations in a container, which got removed later. This can be attached to the same network than the SQL server, so migration is possible. A quick workaround would be writing a script for the `ENTRYPOINT` that first apply migrations and then start the application. But I'm not sure if the last way follows best practices, I think not so much.

**Comment by denDAY04** [comment]
@DMWOO7 Do you have any further information on this issue, e.g. links to where you came to your conclusions? I'm stuck in a similar problem with my current project, wherein I also get this error, with the same framework, when I try scaffolding migrations through the dotnet CLI.

**Comment by DMWOO7** [comment]
Your Dockerfile runs in the _build context_. No networking is avaliable here, since it's designed to install the app and their dependencies. Docker is not designed to do things here, that require network access to special other containers like database or caching servers. So as I said you have two ways to solve this: 

1. Apply your migration before the application starts: That's a quick fix and it works, but brings the side effect that errors of your migration were part of your application start. Depending on your error management/logging this is may not so clean. For example when you're using a CI, those errors may not part of the build process, which could result in false positive jobs (job is successfull, but in fact only the build was okay, the migrations failed)

2. Create a short-lived container, that is started before your application. This container could run in attached mode from a CI, so it's part of your build process. I think that's the smartest solution, since you have a clean seperation of concern: The entrypoint is only responsible for the startup. So if this fails, your app couldn't start, and you don't have to think about migrations, that may failed and cause your app not to start. 

I couldn't find the documents I used in a quick way. The main source was the Docker documentation. For example, the Docker docs follow the mentioned seperation of concern princip. This means, that a container should only have a single task. When we need to do more things in a container, we should outsource this into an additional container.  This is a main principle of docker, which does not only apply in this case. For example, when your app needs a database server, you shouldn't install MySQL and then your app on the same container. Instead, run two linked containers. So it's easy to replace a single peace of the entire application without breaking all the things.

I also found a few questions on SO with related problems. They recommend using a single container, too. But most of them had a lack of explanation, _why_ this is the best way. So I did a analyse on myself to find this out, where this post is a summary of it. Escpecially CI was a big criteria for me, since my Docker containers were deployed using Jenkins. 

For a quick fix, the `ENTRYPOINT` solution can do a good job. Here you simply write a small bash script, which is the entrypoint of the application container. In the script, let EF apply their migrations, maybe with a `dotnet restore` first. Then a simply `dotnet run {YourApp}.dll` starts the app.  For a well designed deployment workflow, I would recommend creating a second container, that is started as part of your deployment for the given reasons. 

This is also a decision of the conept: Fixes like the entrypoint can result into more workarounds. So from time to time, we may end in a maintenance-nightmare, since chancing a single component may break other things. Its also known as the broken-window-theory. Especially in the software development we've this very often: No big additional value is seen in refactoring workarounds, because _they're still working_. Of course they do in the moment. But having too much (old) workarounds can be very expensive. Most notably when they're not well documented.
